tools/uid2dfu.c

```diff
@@ -0,0 +1,161 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Copyright (C) 2015 Simon Newton.
+ */
+
+#include <getopt.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <sysexits.h>
+
+#include "constants.h"
+#include "dfu.h"
+#include "utils.h"
+
+typedef struct {
+  const char *output_file;
+  uint16_t manufacturer_id;
+  uint16_t product_id;
+  uint16_t vendor_id;
+  uint32_t device_id;
+  bool help;
+} Options;
+
+static const char DEFAULT_FILE[] = "uid.dfu";
+
+void DisplayHelpAndExit(const char *arg0, int exit_code) {
+  printf("Usage: %s [options] -m <manufacturer-id> -d <device-id>\n", arg0);
+  printf("  -d, --device <id>  The device ID\n");
+  printf("  -h, --help   Show the help message\n");
+  printf("  -m, --manufacturer <id>  The manufacturer ID\n");
```

Comment:
Are we assuming no-one will ever make production hardware/low volume/one off specials using this then?

Playing devil's advocate, a few interesting questions:
If I buy a unit, will it come with a UID printed on?
What happens if I run the proxy demo code on it, or make a Lab Pack style unit, where do I get the other UIDs from?
Keeping model IDs unique within the manufacturer code (and someone running the code you've written, with set model IDs, against another manufacturer's UID)
General code quality, if the software isn't written by the hardware manufacturer